presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java

```diff
@@ -576,10 +585,47 @@ public int getIndex()
         }
     }
 
+    private RowExpression optimizeExpression(RowExpressionCacheKey cacheKey)
+    {
+        return rowExpressionService.getExpressionOptimizer().optimize(cacheKey.rowExpression, OPTIMIZED, cacheKey.session);
+    }
+
     public enum ColumnMappingKind
     {
         REGULAR,
         PREFILLED,
         INTERIM,
     }
+
+    private static final class RowExpressionCacheKey
+    {
+        private final RowExpression rowExpression;
+        private final ConnectorSession session;
+
+        RowExpressionCacheKey(RowExpression rowExpression, ConnectorSession session)
+        {
+            this.rowExpression = rowExpression;
+            this.session = session;
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(rowExpression, session.getQueryId());
```

Comment:
@bhhari Why not exclude session from hashCode and equals altogether and use identity hashCode and equals on rowExpression?  See https://github.com/prestodb/presto/pull/13872